### ISSUE TITLE
VSTS Task "Service Fabric Application Deployment" : Could not ping any of the provided Service Fabric gateway endpoints

### ISSUE THREAD

**Issue by tale80** [Microsoft/vsts-tasks#7537]
Hi,

Started about 3 weeks ago, all my deployment on Service Fabric failed with a message "**##[error]Could not ping any of the provided Service Fabric gateway endpoints.**" during "**Copying application to image store...**" step.
I deploy from a VSTS 2017 hosted agent on multiple Service Fabric and it's the same result for all.
I tried to increase timeout (but seems to not be applied for this step) in the task, tried the diff package option, revert cluster in older version, but nothing work.

My release definition is a bunch of deploy service fabric app, for each VSTS replace token in the parameter file then deploy. Sometimes the error happen at the first deploy, sometimes, at the last one, sometimes in the middle, i can't find what happen and why.

The only thing i found, is the version of the "**Service Fabric Application Deployment**" used.
the last deployment successfull i have, it was this header in the task :

> ==============================================================================
> 2018-06-06T08:44:32.8059777Z Task         : Service Fabric Application Deployment
> 2018-06-06T08:44:32.8062058Z Description  : Deploy a Service Fabric application to a cluster.
> **2018-06-06T08:44:32.8064320Z Version      : ***.7.9
> 2018-06-06T08:44:32.8066149Z Author       : Microsoft Corporation
> 2018-06-06T08:44:32.8068541Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=8***0***8)
> 2018-06-06T08:44:32.8071185Z ==============================================================================

Now it's 

> ==============================================================================
> 2018-06-20T19:12:29.0249577Z Task         : Service Fabric Application Deployment
> 2018-06-20T19:12:29.0252588Z Description  : Deploy a Service Fabric application to a cluster.
> 2018-06-20T19:12:29.0254786Z Version      : '***.7.***'
> 2018-06-20T19:12:29.0257108Z Author       : Microsoft Corporation
> 2018-06-20T19:12:29.0259591Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=8***0***8)
> 2018-06-20T19:12:29.0262517Z ==============================================================================

In attachment the full log of the error with system.debug on the release.

[log-system.debug.txt](https://github.com/Microsoft/vsts-tasks/files/2123161/log-system.debug.txt)

**Comment by bishal-pdMSFT** [comment]
@tale80, we haven't made any change in task to cause this error. But about 3 weeks backs, SF SDK was updated on Hosted VS2017 agents. Can you please try release on a private agent with an older version of SDK(you can try with 3.0 or older) to see if this is caused by new SDK?
We are also planning to add re-tries if copy fails.

**Comment by tale80** [comment]
@bishal-pdMSFT hi, I will configure my private  agent with older SDK and make a new try next week. I will keep you inform of the result.
Since today i try without compress and without diff package and it seems to be a little better, but still with few errors sometimes so it's not a definitive work around.

**Comment by bishal-pdMSFT** [comment]
@tale80 I would be surprised if disabling compression and diff improves reliability. Infact, I would expect the opposite.
Please do let me know about private agent result.
I tried to repro this error today, but could not get a single failure even in 15 deployments. How big is your application package?

**Comment by tale80** [comment]
@bishal-pdMSFT I splitted my release in multiple small SF packages.
they are about 40/50Mo by app, and i have about 15 packages to deploy one by one.

I would expect the opposite too, but seems to be that. With compression & diff i had errors, without it's ok for now.
I haven't got the time to make the test with a private agent to test an older SDK, i hope i can make the test tomorrow or wednesday

**Comment by bishal-pdMSFT** [comment]
What error do you get while using diff and compression?

**Comment by tale80** [comment]
"##[error]Could not ping any of the provided Service Fabric gateway endpoints." during the "Copying application to image store..." step.
I will make new tests with both options soon to confirm is that the cause of the problem or if i just disable this 2 options in the same time of other operation happenned on hosted agent

**Comment by scary1337** [comment]
Hey Tale80, we are experiencing the same thing, please see my feedback below:

It seems as though we are experiencing transient errors while using VSTS to deploy a service fabric project. We are using a custom powershell script for our staging environment that utilizes similar elements in the pre-rolled powershell script in a fabric project. When we attempt to use the Copy-ServiceFabricApplicationPackage we receive the following error message:

2018-06-25T08:32:12.5368406Z Caught Exception: [System.Fabric.FabricTransientException: Could not ping any of the provided Service Fabric gateway endpoints.

Unfortunately (maybe fortunately), none of our other VSTS deployments are currently doing this (to the same cluster) and only seems to be happening on one specific application type. As this is transient we have built in retry logic, but unfortunately, the copy does not go through. Yesterday (2018-06-24) only one of our umpteen deployments went through.

Is there any advice or properties we could / should be using to avoid this error? We are using HostedVS2017 agents (have not tried with an off-site agent). The command in question (after some modification):

Copy-ServiceFabricApplicationPackage -ApplicationPackagePath $packagePath -ImageStoreConnectionString fabric:ImageStore -ApplicationPackagePathInImageStore $appPathInImageStore -CompressPackage -TimeoutSec 600 -ShowProgress

This works for all of our required deployments except our one service. Any advice / suggestions would be appreciated.

*EDIT*

This only started happening on 2018-06-22 for this specific application type.

*EDIT2*

I have attempted to run an offsite agent now and am receiving the same error. It is starting to look like this is an issue with this applicationType specifically. I unfortunately am not provided with any insight as to how to solve this problem from any documentation. One post mentioned port exhaustion on the FabricClient class, in this instance I am not using that as I am using a custom powershell script.

*EDIT3*

I have now resorted to deploying from my local machine. For some reason my script works from my local machine but NOT the online hosted agents or an offsit agent. My Get-Module returns that my ServiceFabric module is 3.1.0.0, I am not sure if this is reflecting the package correctly.

**Comment by bishal-pdMSFT** [comment]
@scary1337 Are you able to successfully deploy other application types to same cluster?
Hosted agent has SDK version 3.1.283. Can you please try or later this version on your local machine. Please note that this issue is intermittent, so you will have to do multiple deployments to conclude that it works on local machine.

**Comment by scary1337** [comment]
@bishal-pdMSFT - As I stated in my post, all other application types, to the same cluster are working. I am not going to attempt to change my local SDK version as I will not be able to deploy, this is not an option right now. The issue, if truly intermittent, is not ringing true as I am continuously trying to deploy from the online hosted agent with zero success. We have, as stated in the post above, added retries to the `Copy-ServiceFabricApplication` of up to 15 times and the issue still persists, with that application type.

**Comment by tale80** [comment]
@scary1337 i see in your command that you have the "-CompressPackage" option.
For me, it's start to work again when i disabled all optimisation option. Deployments take a little bit more time, but they go to the end.

What application type is? 

I'm currently in release step on my project so all my SF environments are used for test, i can't make a lot of deployment test this week

**Comment by scary1337** [comment]
@tale80 - if I use package compression or not I get the same issue.

Application type being the favric:/MyApplication type. We have several within our cluster and only this specific one seems to be giving an issue. 

Our cluster runs blue green deployments so we have staging and production within the same cluster. Staging uses a custom script to deploy and the production uses the baked in service fabric release with upgrade options.

**Comment by bishal-pdMSFT** [comment]
@scary1337 Are you deploying from your local machine using VSTS? Or are you running the script directly?

**Comment by scary1337** [comment]
@bishal-pdMSFT - Currently I am successfully deploying from my local machine running the script we have created. VSTS, using the same script, is failing (was succeeding before and succeeds on our other services).

**Comment by scary1337** [comment]
@bishal-pdMSFT - that is definitely the case. The SDK versions are different but I can't choose which SDK version gets used on the hosted agents.

**Comment by scary1337** [comment]
Hi @bishal-pdMSFT and @chrisrpatterson - do we have an update on this?

**Comment by bishal-pdMSFT** [comment]
@scary1337 I tried a repro of this issue but it worked for me. Can you think of a special characteristics of this application type package which could be triggering this error? Big overall size? large number of files? large files? etc
@chrisrpatterson is there a way to know if connection to service fabric cluster is being denied/throttled from hosted agent. Please note that the same commands works when tried from local machine.

**Comment by bishal-pdMSFT** [comment]
@scary1337 @tale80 we ran into this issue with another customer. They tried private agent with older sdk and did not see the same failures. I would recommend trying this as this can unblock you. 
You can download last version of SF runtime from [here ](http://download.microsoft.com/download/6/4/4/64446B53-0F81-4E45-9C56-607CA89DBA58/MicrosoftServiceFabric.6.1.456.9494.exe)and sdk from [here](http://download.microsoft.com/download/6/4/4/64446B53-0F81-4E45-9C56-607CA89DBA58/MicrosoftServiceFabricSDK.3.0.456.msi)
Please note that you would need to install both runtime and sdk.

**Comment by bishal-pdMSFT** [comment]
Please do let me know if you see improvement with older sdk.

**Comment by bishal-pdMSFT** [comment]
@scary1337 we have communicated this issue to SF team and they are investigating it.

**Comment by softarch74** [comment]
This issue is hitting us as well. This is a serious hit to our turn around times, as we need to constantly retry deployments.

**Comment by scary1337** [comment]
@softarch74 - it has just hit us in another 2 of our services. @mishal-pdMSFT - surely a rollback of versions on the hosted agents is worthwhile here?

**Comment by scary1337** [comment]
Hi @bishal-pdMSFT  - can we please have a progress update on when this issue is likely to be resolved in production in a newer SDK?

**Comment by bishal-pdMSFT** [comment]
Service fabric team is looking into this issue. I will keep you posted once we get an update.

**Comment by softarch74** [comment]
Is there a serbice fabric github issue we can track?
________________________________

**Comment by bishal-pdMSFT** [comment]
No, there is no github issue. This is being tracked internally. Just to update, we believe we have a root cause . I will update once we start rolling out the fix

**Comment by kuvinodms** [comment]
This fix is under deployment state and will roll out in next few days for all the customers.

**Comment by scary1337** [comment]
hi @kuvinodms - thanks. When would be the best time to test a deployment?

**Comment by softarch74** [comment]
This still is not working. We just got this issue again today.

**Comment by bishal-pdMSFT** [comment]
The fix is still rolling out across scale units. Can you please try again as many of the scale units are done.

**Comment by bishal-pdMSFT** [comment]
The fix has been rolled out completely. Closing this issue. If you still same error, please re-activate this issue.

**Comment by chcosta** [comment]
@bishal-pdMSFT, is there any information about the fix itself that was rolled out to the hosted agents?  We have some static pools of machines that have begun behaving similarly as the issue reported here.  We are seeing this...

```
2018-08-01T17:42:27.1272587Z ##[debug]System.Fabric.FabricTransientException: Could not ping any of the provided Service Fabric gateway endpoints. ---> System.Runtime.InteropServices.COMException: Exception from HRESULT: 0x80071C49
2018-08-01T17:42:27.1285886Z ##[debug]   at System.Fabric.Interop.NativeImageStore.IFabricNativeImageStoreClient3.UploadContent(IntPtr remoteDestination, IntPtr localSource, UInt32 timeoutMilliseconds, FABRIC_IMAGE_STORE_COPY_FLAG copyFlag)
2018-08-01T17:42:27.1298316Z ##[debug]   at System.Fabric.ImageStore.NativeImageStoreClient.UploadContentWrapper(String remoteDestination, String localSource, IImageStoreProgressHandler progressHandler, UInt32 timeoutMilliseconds, CopyFlag imageCopyFlag)
2018-08-01T17:42:27.1310705Z ##[debug]   at System.Fabric.ImageStore.NativeImageStoreClient.<>c__DisplayClass8_0.<UploadContent>b__0()
2018-08-01T17:42:27.1321452Z ##[debug]   at System.Fabric.Interop.Utility.<>c__DisplayClass27_0.<WrapNativeSyncInvoke>b__0()
2018-08-01T17:42:27.1333973Z ##[debug]   at System.Fabric.Interop.Utility.WrapNativeSyncInvoke[TResult](Func`1 func, String functionTag, String functionArgs)
```

Do we need to update to a different VSTS agent version?  to a specific Service Fabric SDK on the machines? Or would you expect that the fix which was implemented would apply uniformly across all vsts agent users that are using the SF task?  Here's the header I'm seeing in the logs...

```
2018-08-02T18:08:28.4214874Z Task         : Service Fabric Application Deployment
2018-08-02T18:08:28.4215176Z Description  : Deploy a Service Fabric application to a cluster.
2018-08-02T18:08:28.4215465Z Version      : 1.7.20
2018-08-02T18:08:28.4215749Z Author       : Microsoft Corporation
2018-08-02T18:08:28.4216077Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=820528)
```

**Comment by chcosta** [comment]
@chrisrpatterson , do you have any additional info on the fix here that I can use to try to resolve our issue with our static pool of machines using Service Fabric?

**Comment by bishal-pdMSFT** [comment]
@chcosta the fix was done in SF task and would be available to all hosted agents. The version of task you are using (1.7.20 as per above log) contains the fix.
Are you sure that this is not an authentication issue? Have you provided correct auth details in service endpoint?

**Comment by bishal-pdMSFT** [comment]
If you feel the issue is happening even with correct service endpoint details, please create a new issue so that it can be tracked properly.

**Comment by tommyja** [comment]
I have this issue. Why am i still seeing version 1.7.19?
2018-08-06T21:30:22.8399720Z Task         : Service Fabric Application Deployment
2018-08-06T21:30:22.8399974Z Description  : Deploy a Service Fabric application to a cluster.
2018-08-06T21:30:22.8400192Z Version      : 1.7.19

**Comment by alalit** [comment]
Hi, we still see this issue and the version comes as 1.7.19. I understand that the fixed version has been rolled out globally, so what is the reason for that?

```
2018-08-06T21:30:22.8399720Z Task : Service Fabric Application Deployment
2018-08-06T21:30:22.8399974Z Description : Deploy a Service Fabric application to a cluster.
2018-08-06T21:30:22.8400192Z Version : 1.7.19
```

**Comment by bishal-pdMSFT** [comment]
Version 1.7.19 contains the fix as well. Please attach debug logs of task.

**Comment by tommyja** [comment]
I think i had the wrong port. It Works now.

**Comment by chcosta** [comment]
Thanks for the details @bishal-pdMSFT.  With no changes on our end, things *appear* to be working again and we're no longer seeing any problems in our Release Service Fabric deployment builds.
